lib/virt_autotest/virtual_network_utils.pm

```diff
@@ -38,7 +38,54 @@ use virt_autotest_base;
 use virt_utils;
 
 our @EXPORT
-  = qw(download_network_cfg prepare_network restore_standalone destroy_standalone restart_libvirtd restart_network restore_guests restore_network destroy_vir_network restore_libvirt_default enable_libvirt_log ssh_setup upload_debug_log check_guest_status);
+  = qw(download_network_cfg prepare_network restore_standalone destroy_standalone restart_libvirtd restart_network restore_guests restore_network
+  destroy_vir_network restore_libvirt_default enable_libvirt_log ssh_setup upload_debug_log check_guest_status save_guest_ip test_network_interface);
+
+sub save_guest_ip {
+    my ($guest, %args) = @_;
+    my $name = $args{name};
+    if (script_run("ping -c3 $guest") != 0) {
+        script_run "sed -i '/$guest/d' /etc/hosts";
+        assert_script_run "virsh domiflist $guest";
+        my $mac_guest = script_output("virsh domiflist $guest | grep $name | grep -oE \"[[:xdigit:]]{2}(:[[:xdigit:]]{2}){5}\"");
+        script_retry "journalctl --no-pager | grep DHCPACK | grep $mac_guest | grep -oE \"([0-9]{1,3}[\.]){3}[0-9]{1,3}\"", delay => 90, retry => 9, timeout => 90;
+        my $gi_guest = script_output("journalctl --no-pager | grep DHCPACK | grep $mac_guest | tail -1 | grep -oE \"([0-9]{1,3}[\.]){3}[0-9]{1,3}\"");
+        assert_script_run "echo '$gi_guest $guest # virtualization' >> /etc/hosts";
+        assert_script_run "ping -c3 $guest";
+    }
+}
+
+sub test_network_interface {
+    my ($guest, %args) = @_;
+    my $net      = $args{net};
+    my $mac      = $args{mac};
+    my $gate     = $args{gate};
+    my $isolated = $args{isolated} // 0;
+    my $target   = $args{target} // script_output("dig +short openqa.suse.de");
+
+    save_guest_ip("$guest", name => $net);
+
+    # Configure the network interface to use DHCP configuration
+    my $nic = script_output "ssh root\@$guest \"grep '$mac' /sys/class/net/*/address | cut -d'/' -f5 | head -n1\"";
+    assert_script_run("ssh root\@$guest \"echo BOOTPROTO=\\'dhcp\\' > /etc/sysconfig/network/ifcfg-$nic\"");
+    if ($guest =~ m/sles11/i) {
+        assert_script_run("ssh root\@$guest service network restart", 90);
+    } else {
+        assert_script_run("ssh root\@$guest systemctl restart wickedd wickedd-dhcp4 wicked", 90);
+    }
```

Comment:
why are we required to restart the network in guest? Is it necessary? I am thinking of adopting this function in my SR-IOV pci passthrough test. In that test, the nic should work immediately after it is plugged into the guest. restarting network may hide some problems.

Reply:
Let me answer this question,  restart network.service as the prepared step in our KVM and Xen guests. During our virtual network test, there just be only one activated network interface in guest system. Moreover, before confirmed each virtual network mode, we need to do some prepared operations. Such as, detach a existed network interface from guest system, and then attach a new network interface into guest system, which modify the existed network configuration for guest system together. So, we'd better to restart network.service before modified the existed network configuration for guest system, imho.